pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,232 @@
+type Manager struct {
+	shutdownGracePeriodRequested    time.Duration
+	shutdownGracePeriodCriticalPods time.Duration
+
+	getPods eviction.ActivePodsFunc
+	killPod eviction.KillPodFunc
+
+	dbusCon               dbusInhibiter
+	nodeShuttingDownNow   bool
+	nodeShuttingDownMutex sync.Mutex
+	inhibitLock           systemd.InhibitLock
+
+	clock clock.Clock
+}
+
+func systemDbus() (dbusInhibiter, error) {
+	bus, err := dbus.SystemBus()
+	if err != nil {
+		return nil, err
+	}
+	return &systemd.DBusCon{SystemBus: bus}, nil
+}
+
+// NewManager returns a new node shutdown manager.
+func NewManager(getPodsFunc eviction.ActivePodsFunc, killPodFunc eviction.KillPodFunc, shutdownGracePeriodRequested, shutdownGracePeriodCriticalPods time.Duration) *Manager {
+	return &Manager{
+		getPods:                         getPodsFunc,
+		killPod:                         killPodFunc,
+		shutdownGracePeriodRequested:    shutdownGracePeriodRequested,
+		shutdownGracePeriodCriticalPods: shutdownGracePeriodCriticalPods,
+		clock:                           clock.RealClock{},
+	}
+}
+
+// Start starts the node shutdown manager and will start watching the node for shutdown events.
+func (m *Manager) Start() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+	if m.shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+
+	systemBus, err := getSystemDbus()
+	if err != nil {
+		return err
+	}
+	m.dbusCon = systemBus
+
+	currentInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+	if err != nil {
+		return err
+	}
+
+	// If the logind's InhibitDelayMaxUSec as configured in (logind.conf) is less than shutdownGracePeriodRequested, attempt to update the value to shutdownGracePeriodRequested.
+	if m.shutdownGracePeriodRequested > currentInhibitDelay {
+		err := m.dbusCon.OverrideInhibitDelay(m.shutdownGracePeriodRequested)
+		if err != nil {
+			return err
+		}
+
+		err = m.dbusCon.ReloadLogindConf()
+		if err != nil {
+			return err
+		}
+
+		// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
+		updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+		if err != nil {
+			return err
+		}
+
+		if updatedInhibitDelay != m.shutdownGracePeriodRequested {
+			return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
+		}
+	}
+
+	err = m.aquireInhibitLock()
+	if err != nil {
+		return err
+	}
+
+	events, err := m.dbusCon.MonitorShutdown()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case isShuttingDown := <-events:
+				klog.V(1).Infof("Shutdown manager detected new shutdown event, isNodeShuttingDownNow: %t", isShuttingDown)
+				m.nodeShuttingDownMutex.Lock()
+				m.nodeShuttingDownNow = isShuttingDown
```

Comment:
so there is a trust that events will be sent properly and there will not be situation when `nodeShuttingDownNow ` was recovered back after already reported as shutting down?

Reply:
should code protect from this happenning?

Reply:
This can happen, e.g. shutdown event (true) event is sent, but then shutdown is cancelled. See comment above.

Reply:
this definitely needs to be explained explicitly in the comment. I would not guess this =)

Reply:
Added a comment here to describe this in more detail.